cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.encryption.guice;
+
+import io.cdap.cdap.security.encryption.AeadCipher;
+import io.cdap.cdap.security.spi.encryption.AeadCipherContext;
+import io.cdap.cdap.security.spi.encryption.CipherException;
+import io.cdap.cdap.security.spi.encryption.CipherInitializationException;
+import java.util.Map;
+
+/**
+ * Default implementation of {@link AeadCipher} which delegates calls to an instance of the SPI
+ * {@link io.cdap.cdap.security.spi.encryption.AeadCipher} with lazy initialization.
+ */
+public class LazyDelegateAeadCipher implements AeadCipher {
+
+  private final io.cdap.cdap.security.spi.encryption.AeadCipher aeadCipher;
+  private final Map<String, String> properties;
+  private Map<String, String> secureProperties;
+  private boolean initialized;
+
+  protected LazyDelegateAeadCipher(io.cdap.cdap.security.spi.encryption.AeadCipher aeadCipher,
+      Map<String, String> properties,
+      Map<String, String> secureProperties) {
+    this.aeadCipher = aeadCipher;
+    this.properties = properties;
+    this.secureProperties = secureProperties;
+    this.initialized = false;
+  }
+
+
+  private void initialize() throws CipherInitializationException {
+    aeadCipher.initialize(new AeadCipherContext(properties, secureProperties));
+    // Help garbage collect secure properties to avoid keeping them in memory.
+    secureProperties = null;
+    this.initialized = true;
+  }
+
+  @Override
+  public byte[] encrypt(byte[] plainData, byte[] associatedData) throws CipherException {
+    if (!initialized) {
```

Comment:
This would not be concurrent safe. It's better to make it safe by using a proper double sync singleton technique.

Reply:
Good catch, done